src/daemon/rpmostreed-transaction-types.cxx

```diff
@@ -2402,6 +2402,115 @@ rpmostreed_transaction_new_refresh_md (GDBusMethodInvocation *invocation,
 
   return (RpmostreedTransaction *) self;
 
+}
+/* =================================== Search ==================================== */
+
+typedef struct {
+  RpmostreedTransaction parent;
+  char *osname;
+} SearchTransaction;
+
+typedef RpmostreedTransactionClass SearchTransactionClass;
+
+GType search_transaction_get_type (void);
+
+G_DEFINE_TYPE (SearchTransaction,
+               search_transaction,
+               RPMOSTREED_TYPE_TRANSACTION)
+
+static void
+search_transaction_finalize (GObject *object)
+{
+  SearchTransaction *self;
+
+  self = (SearchTransaction *) object;
+  g_free (self->osname);
+
+  G_OBJECT_CLASS (search_transaction_parent_class)->finalize (object);
+}
+
+static gboolean
+search_transaction_execute (RpmostreedTransaction *transaction,
+                            GCancellable *cancellable,
+                            GError **error)
+{
+  SearchTransaction *self = (SearchTransaction *) transaction;
+  OstreeSysroot *sysroot = rpmostreed_transaction_get_sysroot (transaction);
+
+  g_autoptr(GString) title = g_string_new ("search");
+
+  g_autoptr(OstreeDeployment) cfg_merge_deployment =
+    ostree_sysroot_get_merge_deployment (sysroot, self->osname);
+  g_autoptr(OstreeDeployment) origin_merge_deployment =
+    rpmostree_syscore_get_origin_merge_deployment (sysroot, self->osname);
+
+  /* but set the source root to be the origin merge deployment's so we pick up releasever */
+  g_autofree char *origin_deployment_root =
+    rpmostree_get_deployment_root (sysroot, origin_merge_deployment);
+
+  OstreeRepo *repo = ostree_sysroot_repo (sysroot);
+  g_autoptr(RpmOstreeContext) ctx = rpmostree_context_new_client (repo);
+
+  /* We could bypass rpmostree_context_setup() here and call dnf_context_setup() ourselves
+   * since we're not actually going to perform any installation. Though it does provide us
+   * with the right semantics for install/source_root. */
+  if (!rpmostree_context_setup (ctx, NULL, origin_deployment_root, cancellable, error))
+    return FALSE;
+
+  if (force)
+    rpmostree_context_set_dnf_caching (ctx, RPMOSTREE_CONTEXT_DNF_CACHE_NEVER);
+
+  /* point libdnf to our repos dir */
+  rpmostree_context_configure_from_deployment (ctx, sysroot, cfg_merge_deployment);
+
+  /* don't even bother loading the rpmdb */
+  if (!rpmostree_context_download_metadata (ctx, DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB,
+                                            cancellable, error))
+    return FALSE;
+
```

Comment:
And here (but again, in `osexperimental_handle_search`) is where you'd then use the `HyQuery` API to search for the patterns passed in. Though actually, you can probably use `rpmostree_get_matching_packages`.